internal/pkg/composable/providers/kubernetessecrets/kubernetes_secrets.go

```diff
@@ -46,67 +55,223 @@ func ContextProviderBuilder(logger *logger.Logger, c *config.Config, managed boo
-	// key = "kubernetes_secrets.somenamespace.somesecret.value"
+	if p.config.DisableCache {
+		valid := p.validateKey(key)
+		if valid {
+			return p.fetchSecretWithTimeout(key)
+		} else {
+			return "", false
+		}
+	} else {
+		return p.getFromCache(key)
+	}
+}
+
+// Run initializes the k8s secrets context provider.
+func (p *contextProviderK8sSecrets) Run(ctx context.Context, comm corecomp.ContextProviderComm) error {
+	client, err := getK8sClientFunc(p.config.KubeConfig, p.config.KubeClientOptions)
+	if err != nil {
+		p.logger.Debugf("kubernetes_secrets provider skipped, unable to connect: %s", err)
+		return nil
+	}
 	p.clientMx.Lock()
-	client := p.client
+	p.client = client
 	p.clientMx.Unlock()
-	if client == nil {
-		return "", false
+
+	if !p.config.DisableCache {
+		go p.updateSecrets(ctx)
+	}
+
+	<-comm.Done()
+
+	p.clientMx.Lock()
+	p.client = nil
+	p.clientMx.Unlock()
+	return comm.Err()
+}
+
+func getK8sClient(kubeconfig string, opt kubernetes.KubeClientOptions) (k8sclient.Interface, error) {
+	return kubernetes.GetKubernetesClient(kubeconfig, opt)
+}
+
+// Update the secrets in the cache every RefreshInterval
+func (p *contextProviderK8sSecrets) updateSecrets(ctx context.Context) {
+	timer := time.NewTimer(p.config.RefreshInterval)
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-timer.C:
+			p.updateCache()
+			timer.Reset(p.config.RefreshInterval)
+		}
+	}
+}
+
+// mergeWithCurrent merges the updated map with the cache map.
+// This function needs to be called between the mutex lock for the map.
+func (p *contextProviderK8sSecrets) mergeWithCurrent(updatedMap map[string]*secretsData) map[string]*secretsData {
+	merged := make(map[string]*secretsData)
+	for name, data := range p.secretsCache {
+		diff := time.Since(data.lastAccess)
+		if diff < p.config.TTLDelete {
+			merged[name] = data
+		}
+	}
+	for name, data := range updatedMap {
+		merged[name] = data
+	}
+	return merged
+}
+
+func (p *contextProviderK8sSecrets) updateCache() {
+	// deleting entries does not free the memory, so we need to create a new map
+	// to place the secrets we want to keep
+	cacheTmp := make(map[string]*secretsData)
+
+	p.secretsCacheMx.RLock()
+	for name, data := range p.secretsCache {
+		diff := time.Since(data.lastAccess)
+		if diff < p.config.TTLDelete {
+			value, ok := p.fetchSecretWithTimeout(name)
+			if ok {
+				newData := &secretsData{
+					value:      value,
+					lastAccess: data.lastAccess,
+				}
+				cacheTmp[name] = newData
+			}
+
+		}
+	}
+	p.secretsCacheMx.RUnlock()
+
+	// While the cache was updated, it is possible that some secret was added through another go routine.
+	// We need to merge the updated map with the current cache map to catch the new entries and avoid
+	// loss of data.
+	p.secretsCacheMx.Lock()
+	p.secretsCache = p.mergeWithCurrent(cacheTmp)
+	p.secretsCacheMx.Unlock()
+}
+
+func (p *contextProviderK8sSecrets) getFromCache(key string) (string, bool) {
+	p.secretsCacheMx.Lock()
```

Comment:
I believe now RLock and RUnlock should be used.

Reply:
Not possible, because we are calling `addToCache` in this function, and we need that lock

Reply:
TBH I don't understand why we need to hold the lock while we are waiting to fetchSecretWithTimeout

Reply:
We are adding a value to the cache, and then we are returning the value that is inside the cache. If we don't hold the lock there, we get a racing condition from the tests.

Reply:
Ok can you explain to me what is wrong with this code? I am not holding the lock while waiting for 
fetchSecretWithTimeout. 

```
func (p *contextProviderK8sSecrets) getFromCache(key string) (string, bool) {
	p.secretsCacheMx.RLock()
	defer p.secretsCacheMx.RUnlock()

	data, ok := p.secretsCache[key]
	if ok {
		data.lastAccess = time.Now()
		pass := data.value
		return pass, ok
	}
	data, ok = p.addToCache(key)
	// if it was not possible to fetch the secret, return
	return data.value, ok
}

func (p *contextProviderK8sSecrets) addToCache(key string) (*secretsData, bool) {
	valid := p.validateKey(key)
	if !valid {
		return &secretsData{
			value: "",
		}, false
	}

	value, ok := p.fetchSecretWithTimeout(key)
	data := &secretsData{
		value: value,
	}
	if ok {
		p.secretsCacheMx.Lock()
		p.secretsCache[key] = data
		p.secretsCacheMx.Unlock()
	}
	return data, ok
}
```

Reply:
Yes because of the defer wich will keep the read lock while adToCache runs. Actually what it needs is to just remove the read lock after reading

```
func (p *contextProviderK8sSecrets) getFromCache(key string) (string, bool) {
	p.secretsCacheMx.RLock()
	data, ok := p.secretsCache[key]
	p.secretsCacheMx.RUnlock()
	if ok {
		data.lastAccess = time.Now()
		pass := data.value
		return pass, ok
	}
	data, ok = p.addToCache(key)
	return data.value, ok
}
```
One of the tests is not passing though ...

Reply:
From my side it's good to go! Thanks for patience!